Signal/src/util/OWSBackupExport.m

```diff
@@ -0,0 +1,601 @@
+
+            // Copy attachments.
+            [srcTransaction
+                enumerateKeysAndObjectsInCollection:[TSAttachmentStream collection]
+                                         usingBlock:^(NSString *key, id object, BOOL *stop) {
+                                             if (self.isComplete) {
+                                                 *stop = YES;
+                                                 return;
+                                             }
+                                             if (![object isKindOfClass:[TSAttachment class]]) {
+                                                 OWSProdLogAndFail(
+                                                     @"%@ unexpected class: %@", self.logTag, [object class]);
+                                                 return;
+                                             }
+                                             if ([object isKindOfClass:[TSAttachmentStream class]]) {
+                                                 TSAttachmentStream *attachmentStream = object;
+                                                 NSString *_Nullable filePath = attachmentStream.filePath;
+                                                 if (filePath) {
+                                                     OWSAssert(attachmentStream.uniqueId.length > 0);
+                                                     self.attachmentFilePathMap[attachmentStream.uniqueId] = filePath;
+                                                 }
+                                             }
+                                             TSAttachment *attachment = object;
+                                             [attachment saveWithTransaction:dstTransaction];
+                                             copiedAttachments++;
+                                             copiedEntities++;
+                                         }];
+        }];
+    }];
+
+    // TODO: Should we do a database checkpoint?
+
+    DDLogInfo(@"%@ copiedThreads: %llu", self.logTag, copiedThreads);
+    DDLogInfo(@"%@ copiedMessages: %llu", self.logTag, copiedInteractions);
+    DDLogInfo(@"%@ copiedEntities: %llu", self.logTag, copiedEntities);
+    DDLogInfo(@"%@ copiedAttachments: %llu", self.logTag, copiedAttachments);
+
+    [self.backupStorage logFileSizes];
+
+    // Capture the list of files to save.
+    self.databaseFilePaths = [@[
+        self.backupStorage.databaseFilePath,
+        self.backupStorage.databaseFilePath_WAL,
+        self.backupStorage.databaseFilePath_SHM,
+    ] mutableCopy];
+
+    // Close the database.
+    self.dstDBConnection = nil;
+    self.backupStorage = nil;
+
+    return YES;
+}
+
+- (void)saveToCloud:(OWSBackupExportCompletion)completion
+{
+    DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    self.databaseRecordMap = [NSMutableDictionary new];
+    self.attachmentRecordMap = [NSMutableDictionary new];
+
+    [self saveNextFileToCloud:completion];
```

Comment:
The actual process of saving content to the cloud is super async.  Every time we succeed (or ignore a failure), we call `saveNextFileToCloud` which tries to start the next upload or completes the save process.

Reply:
Added this question to the design doc.  

* We certainly want "best performance possible" when restoring a backup, since that's a (probably?) blocking operation.  Apple's CloudKit docs stress that they've put a lot of effort into optimizing their network utilization, so when doing a restore, I just ask for everything at once and let CloudKit figure out how to perform the downloads.
* When doing "backup exports", I'm less clear on what the best approach is.  For one thing, I don't want backups to interfere with user activity, e.g. sending and receiving messages.